### PR TITLE
(maint) Add environment convergence docs

### PR DESCRIPTION
**Pull request by joshcooper**
Describe agent vs server specified environments, strict_environment_mode
and use_last_environment.

### CODE REVIEW

File: docs/environment_convergence.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+PE ships with a `classifier` node terminus that communicates with the PE classifier via REST. So by default, PE uses server-specified environments.
+
+However, open source defaults to a `plain` node terminus, which means it falls back to agent-specified environments.
+
+When the agent negotiates with the server, the server may respond in one of the following ways:
+
+  1. Tell the agent to switch to the server-specified environment.
+  1. Allow the agent to continue using its current agent-specified environment.
+  1. Fail the run due to a classification conflict. Conflicts can arise if the agent is assigned to multiple environments at the same time, as can happen if the agent's facts match multiple rules in the classifier.
+  1. Fail the request because the agent's requested environment doesn't exist on the server.
```
Comment (cthorn42) [inline docs/environment_convergence.md:24]: I'm not sure if this is a mark down issue or not, but this numbering looks odd.

Comment (joshcooper) [reply docs/environment_convergence.md]: I thought this was a standard markdown thing, but maybe it's a github flavored markdown thing so it autoincrements. In any case, I changed it to 1., 2., etc

File: docs/environment_convergence.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+When running in an agent-specified context, i.e. the server **allows** the agent to decide, then the agent's run may result in one of the following outcomes (again ignoring networking issues):
+
+1. If the environment exists on the server, then the agent will use that for the duration of the run, even if it needs to download facts and plugins from the server.
+1. If the environment doesn't exist on the server:
+  1. By default, switch to the server-specified environment, typically the value of `Puppet[:environment]` on the server.
+  1. Otherwise, if `strict_enviroment_mode` is enabled, then fail the run, because the user wants to strictly use the requested environment.
+
+## Last Used Environment
+
+Older versions of the agent used to make a node request at the beginning of the run to determine which environment to start off in. Newer versions do not and instead cache the last used environment in `Puppet[:lastrunfile]`. Doing so eliminates the agent's node request and several requests among server, classifier, puppetdb and postgres.
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply docs/environment_convergence.md]: Updated

File: docs/environment_convergence.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+For security reasons, puppet defaults to **server-specified environments**. This means the server always decides which environment to assign to each agent. This is important because we don't want an agent to request a catalog from an arbitrary environment, as the server might include a class containing sensitive data.
+
+However, there are cases during [iterative development](https://puppet.com/docs/pe/latest/environment_based_testing.html) where it is necessary for a trusted user to be able to run the agent against an environment contained in a feature branch. For example, `puppet agent -t --environment <feature>`. If the code works as intended, then it provides confidence that the feature branch code can be merged. We refer to this workflow as an **agent-specified environment**.
+
+## Agent and Server Negotiation
+
+The server decides which environment an agent should use by asking its node terminus to classify the node.Terminus just means there's an implementation of an interface that knows how to perform CRUD operations on `Puppet::Node` objects. The terminus is responsible for returning a node object, including the environment that the node is assigned to.
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply docs/environment_convergence.md]: Updated

**Review by cthorn42 — APPROVED** [review]

**Pull request merged**